lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+                GetBoolValue(constCase->thing).value(), caseStmt.source);
+          }
+        } else {
+          if (selectCaseStmtType.value().category() ==
+              TypeCategory::Logical) {  // C1148 (R1140)
+            context_.Say(caseStmt.source,
+                "SELECT CASE expression of type LOGICAL must not have range of case value"_err_en_US);
+            continue;
+          }
+          const auto &rangeCase{
+              std::get<parser::CaseValueRange::Range>(caseValues.u)};
+          const auto &lower{rangeCase.lower};
+          if (lower &&
+              !IsValidSelectCaseType(selectCaseStmtType.value(),
+                  lower.value().thing, caseStmt.source)) {
+            continue;
+          }
+
+          const auto &upper{rangeCase.upper};
+          if (upper &&
+              !IsValidSelectCaseType(selectCaseStmtType.value(),
+                  upper.value().thing, caseStmt.source)) {
+            continue;
+          }
+
+          if (selectCaseStmtType.value().category() == TypeCategory::Integer) {
+            const auto lowerVal = lower
+                ? std::optional(GetIntValue(lower->thing).value())
+                : std::nullopt;
+            const auto upperVal = upper
+                ? std::optional(GetIntValue(upper->thing).value())
+                : std::nullopt;
+            selectCaseStmts.Insert(lowerVal, upperVal, caseStmt.source);
+          } else {  // TypeCategory::Character
+            const auto lowerVal = lower
+                ? std::optional(GetStringValue(lower->thing).value())
+                : std::nullopt;
+            const auto upperVal = upper
+                ? std::optional(GetStringValue(upper->thing).value())
+                : std::nullopt;
+            selectCaseStmts.Insert(lowerVal, upperVal, caseStmt.source);
+          }
+        }
+      }
+    }
+  }
+}
+
+bool SelectStmtChecker::IsValidSelectCaseType(
+    const evaluate::DynamicType &expectedType,
+    const parser::ConstantExpr &constCase, const parser::CharBlock &src) {
+  if (const auto *caseval{GetExpr(constCase)}) {
+    if (auto type{caseval->GetType()}) {
+      if (type->category() != expectedType.category()) {  // C1147 (R1140)
+        if (expectedType.category() == TypeCategory::Integer) {
+          context_.Say(
+              src, "SELECT CASE value must be of type INTEGER"_err_en_US);
+        } else if (expectedType.category() == TypeCategory::Character) {
+          context_.Say(
+              src, "SELECT CASE value must be of type CHARACTER"_err_en_US);
+        } else {  // TypeCategory::Logical
+          context_.Say(
+              src, "SELECT CASE value must be of type LOGICAL"_err_en_US);
+        }
+      } else if (expectedType.category() == TypeCategory::Character) {
+        if (expectedType.kind() != type->kind()) {
+          context_.Say(src,
+              "SELECT CASE value kind must be same as SELECT CASE expression kind"_err_en_US);
```

Comment:
You could say what kind it is supposed to be. Like above where you say what type it is supposed to be.

Reply:
I will modify to display the expected kind for character.